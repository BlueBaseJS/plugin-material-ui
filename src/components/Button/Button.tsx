import {
	ButtonDefaultProps,
	ButtonProps
} from '@bluebase/components';
import MUIButton from '@material-ui/core/Button';
import React from 'react';
import { Theme } from '@bluebase/core';
import { objectMapper } from '@bluebase/component-mapper';
import { withPropsStyles } from '../../withPropsStyles';

const styles = ({ color }: any, theme: Theme) => {

	switch (color) {
		case 'primary':
			return {};
		case 'secondary':
			return {};
		case 'success':
			return ({
				contained: {
					backgroundColor: theme.palette.success.main,
					color: theme.palette.success.contrastText,
					// tslint:disable-next-line: object-literal-sort-keys
					'&:hover': {
						backgroundColor: theme.palette.success.dark,
						color: theme.palette.success.contrastText,
					},
				},
				outlined: {
					borderColor: theme.palette.success.main,
					color: theme.palette.success.main,
					// tslint:disable-next-line: object-literal-sort-keys
					'&:hover': {
						backgroundColor: theme.palette.success.contrastText,
					},
				},
				root: {
					color: theme.palette.success.main,
				},
			});
		case 'error':

			return ({
				contained: {
					backgroundColor: theme.palette.error.main,
					color: theme.palette.error.contrastText,
					// tslint:disable-next-line: object-literal-sort-keys
					'&:hover': {
						backgroundColor: theme.palette.error.dark,
						color: theme.palette.error.contrastText,
					},
				},
				outlined: {
					borderColor: theme.palette.error.main,
					color: theme.palette.error.main,
					// tslint:disable-next-line: object-literal-sort-keys
					'&:hover': {
						backgroundColor: theme.palette.error.contrastText,
					},
				},
				root: {
					color: theme.palette.error.main,
				}
			});
		case 'warning':
			return ({
				contained: {
					backgroundColor: theme.palette.warning.main,
					color: theme.palette.warning.contrastText,
					// tslint:disable-next-line: object-literal-sort-keys
					'&:hover': {
						backgroundColor: theme.palette.warning.dark,
						color: theme.palette.warning.contrastText,
					},
				},
				outlined: {
					borderColor: theme.palette.warning.main,
					color: theme.palette.warning.main,
					// tslint:disable-next-line: object-literal-sort-keys
					'&:hover': {
						backgroundColor: theme.palette.background.default,
					},
				},
				root: {
					color: theme.palette.warning.main,
				}
			});
		default:
			return null;
	}


};
const map = {
	// If color is primary, secondary or default set as is
	color: ({ color }: ButtonProps) => {
		if (color === 'primary' || color === 'secondary' || color === 'default') {
			return color;
		}

		return;
	},

	onChange: ({ onChange, onValueChange }: any) => (event: any, checked: boolean) => {
		if (onChange) {
			onChange(event, checked);
		}

		if (onValueChange) {
			onValueChange(event.target.value, checked);
		}
	}

};

export const Button = withPropsStyles(styles)((props: any) => {

	const newProps = objectMapper(props, map, { rest: true, ignore: ['onValueChange'] });

	const { loading, color, children, title, classes, indeterminate, ...common } = newProps;

	if (Object.keys(classes).length > 0) {
		delete common.color;
	}


	return (
		<MUIButton style={classes} {...props}>

			{props.children ? props.children : props.title}
		</MUIButton>
	);
}) as React.ComponentType<ButtonProps>;

Button.defaultProps = ButtonDefaultProps;