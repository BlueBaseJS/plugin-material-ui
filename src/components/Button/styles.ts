import { SwitchProps } from '@bluebase/components';
import { Theme } from '@bluebase/core';
import { Theme as MuiTheme } from '@material-ui/core';
import { fade } from '@material-ui/core/styles/colorManipulator';

export const styles = ({ color }: SwitchProps, muiTheme: MuiTheme, theme: Theme) => {
	if (!color || color === 'default') {
		return;
	}

	const colors: {
		hover?: string;
		main?: string;
		text?: string;
	} = {};

	// If color is NOT primary, secondary or default then create custom styles
	if (
		color === 'primary' ||
		color === 'secondary' ||
		color === 'success' ||
		color === 'error' ||
		color === 'warning'
	) {
		colors.hover = (theme.palette as any)[color].dark;
		colors.main = (theme.palette as any)[color].main;
		colors.text = (theme.palette as any)[color].contrastText;
	} else {
		colors.hover = color;
		colors.main = color;
		colors.text = muiTheme.palette.getContrastText(color);
	}

	return {
		contained: {},
		outlined: {},
		// text: {},

		root: {
			'&$contained': {
				backgroundColor: colors.main,
				color: colors.text,

				'&:hover': {
					backgroundColor: colors.hover,
				},
			},
			/* Styles applied to the root element if `variant="outlined"` and `color="primary"`. */
			'&$outlined': {
				'&:hover': {
					'@media (hover: none)': {
						backgroundColor: 'transparent',
					},
					backgroundColor: fade(colors.main as string, theme.palette.action.hoverOpacity),
					border: `1px solid ${colors.main as string}`,
					// Reset on touch devices, it doesn't add specificity
				},
				border: `1px solid ${fade(colors.main as string, 0.5)}`,
				color: colors.main,
			},
			'&$text': {
				'&:hover': {
					'@media (hover: none)': {
						backgroundColor: 'transparent',
					},
					backgroundColor: fade(colors.main as string, theme.palette.action.hoverOpacity),
					// Reset on touch devices, it doesn't add specificity
				},
				color: colors.main,
			},
		},
		/* Styles applied to the root element if `variant="text"` */
		text: {},
	};
};
