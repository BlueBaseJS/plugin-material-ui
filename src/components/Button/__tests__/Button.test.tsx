import { Button } from '../Button';
import React from 'react';
import { Text } from 'react-native';
import { mount,  shallow } from 'enzyme';
import Plugin from '../../../index';
import { BlueBaseApp } from '@bluebase/core';


test('Button component should use title prop to show children', () => {
	const component = mount(
		<BlueBaseApp plugins={[Plugin]}>
			<Button title="Foo" />
		</BlueBaseApp>
	);
	expect(component).toMatchSnapshot();
	expect(component.find('Button').prop('title')).toEqual('Foo');
});

test('Button component should use children prop to show content', () => {
	const component = shallow(
		<Button title="Foo">
			<Text>Bar</Text>
		</Button>
	);
	// expect(component).toMatchSnapshot();
	expect(component.find('Text').text()).toEqual('Bar');
});
