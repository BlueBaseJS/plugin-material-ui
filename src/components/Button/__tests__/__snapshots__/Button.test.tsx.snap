// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should use title prop to show children 1`] = `
<BlueBaseApp
  plugins={
    Array [
      Object {
        "categories": Array [
          "ui",
        ],
        "defaultConfigs": Object {},
        "description": "Material UI (web) comes to BlueBase!",
        "enabled": true,
        "key": "material-ui",
        "name": "Material UI",
        "value": Object {
          "assets": Object {},
          "components": Object {
            "Avatar": [Function],
            "BottomNavigation": [Function],
            "BottomNavigationAction": [Function],
            "Button": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "color": "primary",
                "fullWidth": false,
                "loading": false,
                "size": "medium",
                "variant": "contained",
              },
              "render": [Function],
            },
            "Card": [Function],
            "CardActions": [Function],
            "CardContent": [Function],
            "CardCover": [Function],
            "CardHeader": [Function],
            "Checkbox": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Dialog": [Function],
            "DialogActions": [Function],
            "DialogContent": [Function],
            "DialogTitle": [Function],
            "Divider": [Function],
            "DrawerActions": [Function],
            "DrawerItem": [Function],
            "DrawerLayout": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "DrawerSection": [Function],
            "IconButton": [Function],
            "List": [Function],
            "ListAvatar": [Function],
            "ListIcon": [Function],
            "ListItem": [Function],
            "ListSubheader": [Function],
            "Menu": [Function],
            "MenuItem": [Function],
            "Picker": [Function],
            "PickerItem": [Function],
            "Radio": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "RadioGroup": [Function],
            "Slider": [Function],
            "Switch": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Tab": [Function],
            "Tabs": [Function],
            "TextInput": [Function],
          },
          "filters": Object {
            "bluebase.boot.end": [Function],
          },
          "fonts": Object {},
          "routes": undefined,
          "themes": Object {},
        },
        "version": "0.0.1",
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text>
        <Text>
          Loading
        </Text>
      </Text>
    </View>
  </View>
</BlueBaseApp>
`;
